File: packages/agent/src/adapters/codex-app-server/spawn.ts
import { type ChildProcess, spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { delimiter, dirname } from "node:path";
import type { Readable, Writable } from "node:stream";
import type { ProcessSpawnedCallback } from "../../types";
import { Logger } from "../../utils/logger";
import { stripElectronNodeShimFromPath } from "../../utils/spawn-env";
import { CodexSettingsManager } from "../codex/settings";

export interface CodexAppServerProcessOptions {
  /** Path to the native `codex` CLI binary (the one that exposes `app-server`). */
  binaryPath: string;
  cwd?: string;
  apiBaseUrl?: string;
  apiKey?: string;
  /**
   * Private CODEX_HOME for this run (skills + config). Without it codex falls
   * back to the user's ~/.codex, whose ambient plugins/MCP servers can stall
   * every turn (a broken plugin MCP blocks turn/start for its full timeout).
   */
  codexHome?: string;
  /** Guidance appended to Codex's base prompt via `developer_instructions`. */
  developerInstructions?: string;
  /** Extra codex `-c key=value` config overrides (e.g. auto_compact_token_limit). */
  configOverrides?: Record<string, string | number>;
  logger?: Logger;
  processCallbacks?: ProcessSpawnedCallback;
}

export interface CodexAppServerProcess {
  process: ChildProcess;
  stdin: Writable;
  stdout: Readable;
  kill: () => void;
}

export function buildAppServerArgs(
  options: CodexAppServerProcessOptions,
): string[] {
  const args: string[] = ["app-server"];

  args.push("-c", "features.remote_models=false");

  // Ambient plugins from the user's config.toml inject MCP servers and
  // session-start hooks into PostHog sessions (e.g. an unauthenticated plugin
  // MCP failing every thread, hooks wedging turns). Threads only get the MCP
  // servers PostHog injects, so disable the plugin system outright.
  args.push("-c", "features.plugins=false");

  // OS sandbox gated on platform (= availability): macOS Seatbelt → workspace-write
  // (keeps the sandbox engaged so a per-turn readOnly can tighten it and block
  // edits); linux/windows have no sandbox launcher and would panic, so
  // danger-full-access (the enclosing docker/Modal sandbox isolates instead).
  args.push(
    "-c",
    process.platform === "darwin"
      ? `sandbox_mode="workspace-write"`
      : `sandbox_mode="danger-full-access"`,
  );

  // Disable the user's ambient ~/.codex MCP servers so the adapter only exposes
  // MCP servers PostHog injects per-thread; otherwise codex fails connecting to them.
  for (const name of new CodexSettingsManager(
    options.cwd ?? process.cwd(),
  ).getSettings().mcpServerNames) {
    // codex's `-c` parser rejects quoted/special key segments; skip such names.
    if (!/^[A-Za-z0-9_-]+$/.test(name)) continue;
    args.push("-c", `mcp_servers.${name}.enabled=false`);
  }

  if (options.apiBaseUrl) {
    args.push("-c", `model_provider="posthog"`);
    args.push("-c", `model_providers.posthog.name="PostHog Gateway"`);
    args.push("-c", `model_providers.posthog.base_url="${options.apiBaseUrl}"`);
    args.push("-c", `model_providers.posthog.wire_api="responses"`);
    args.push(
      "-c",
      `model_providers.posthog.env_key="POSTHOG_GATEWAY_API_KEY"`,
    );
  }

  // developer_instructions are set per-thread in thread/start (with the host's
  // task system prompt), not as a spawn-level global default.

  // Numbers/bools go bare; strings are quoted, matching codex's `-c` parser.
  for (const [key, value] of Object.entries(options.configOverrides ?? {})) {
    args.push(
      "-c",
      `${key}=${typeof value === "number" ? value : `"${value}"`}`,
    );
  }

  return args;
}

export function spawnCodexAppServerProcess(
  options: CodexAppServerProcessOptions,
): CodexAppServerProcess {
  const logger =
    options.logger ?? new Logger({ debug: true, prefix: "[CodexAppServer]" });

  if (!existsSync(options.binaryPath)) {
    throw new Error(
      `codex binary not found at ${options.binaryPath}. Run "node apps/code/scripts/download-binaries.mjs" to download it.`,
    );
  }

  const env: NodeJS.ProcessEnv = { ...process.env };
  delete env.ELECTRON_RUN_AS_NODE;
  delete env.ELECTRON_NO_ASAR;
  if (options.apiKey) {
    env.POSTHOG_GATEWAY_API_KEY = options.apiKey;
  }
  if (options.codexHome) {
    env.CODEX_HOME = options.codexHome;
  }
  env.PATH = `${dirname(options.binaryPath)}${delimiter}${stripElectronNodeShimFromPath(env.PATH) ?? ""}`;

  const args = buildAppServerArgs(options);

  logger.info("Spawning codex app-server process", {
    command: options.binaryPath,
    args,
    cwd: options.cwd,
  });

  const child = spawn(options.binaryPath, args, {
    cwd: options.cwd,
    env,
    stdio: ["pipe", "pipe", "pipe"],
    detached: process.platform !== "win32",
  });

  child.stderr?.on("data", (data: Buffer) => {
    logger.warn("codex app-server stderr:", data.toString());
  });

  child.on("error", (err) => {
    logger.error("codex app-server process error:", err);
  });

  child.on("exit", (code, signal) => {
    logger.info("codex app-server process exited", { code, signal });
    if (child.pid && options.processCallbacks?.onProcessExited) {
      options.processCallbacks.onProcessExited(child.pid);
    }
  });

  if (!child.stdin || !child.stdout) {
    throw new Error(
      "Failed to get stdio streams from codex app-server process",
    );
  }

  if (child.pid && options.processCallbacks?.onProcessSpawned) {
    options.processCallbacks.onProcessSpawned({
      pid: child.pid,
      command: options.binaryPath,
    });
  }

  return {
    process: child,
    stdin: child.stdin,
    stdout: child.stdout,
    kill: () => {
      logger.info("Killing codex app-server process", { pid: child.pid });
      child.stdin?.destroy();
      child.stdout?.destroy();
      child.stderr?.destroy();
      child.kill("SIGTERM");
    },
  };
}

File: rust/crates/codex-driver/src/spawn.rs
//! codex app-server process spawn: argument construction, environment, and
//! binary resolution. Port of `spawn.ts` (`buildAppServerArgs` /
//! `spawnCodexAppServerProcess`) and `binary-path.ts`, for the sandbox
//! layout (launch cwd `/scripts`, binary vendored by `@openai/codex`).

use std::path::PathBuf;
use std::process::Stdio;

use tokio::process::{Child, ChildStdin, ChildStdout, Command};

#[derive(Debug, Clone, Default)]
pub struct AppServerOptions {
    pub binary_path: PathBuf,
    pub cwd: String,
    /// Gateway base URL; configures the `posthog` model provider when set.
    pub api_base_url: Option<String>,
    /// Gateway key, exported as `POSTHOG_GATEWAY_API_KEY`.
    pub api_key: Option<String>,
    /// Private CODEX_HOME for this run; without it codex uses ~/.codex.
    pub codex_home: Option<String>,
}

/// Resolve the native codex binary: `POSTHOG_CODEX_BINARY_PATH` env override,
/// else the binary vendored by `@openai/codex`'s platform sub-package under
/// the working directory's node_modules (the sandbox installs `@posthog/agent`
/// with its `@openai/codex` dependency into /scripts).
pub fn resolve_codex_binary() -> Option<PathBuf> {
    if let Ok(explicit) = std::env::var("POSTHOG_CODEX_BINARY_PATH") {
        if !explicit.is_empty() {
            return Some(PathBuf::from(explicit));
        }
    }
    let (pkg, triple) = if cfg!(target_arch = "aarch64") {
        ("@openai/codex-linux-arm64", "aarch64-unknown-linux-musl")
    } else {
        ("@openai/codex-linux-x64", "x86_64-unknown-linux-musl")
    };
    let vendored = PathBuf::from(format!("./node_modules/{pkg}/vendor/{triple}/bin/codex"));
    if vendored.exists() {
        return Some(vendored);
    }
    None
}

/// Port of `buildAppServerArgs` for the cloud path. The macOS Seatbelt branch
/// and the ambient-config MCP disabling are omitted: the driver only runs in
/// linux sandboxes, where there is no OS sandbox launcher (the enclosing
/// docker/Modal sandbox isolates instead) and CODEX_HOME is image-controlled.
pub fn build_app_server_args(options: &AppServerOptions) -> Vec<String> {
    let mut args: Vec<String> = vec!["app-server".into()];

    let mut push_config = |value: &str| {
        args.push("-c".into());
        args.push(value.to_string());
    };

    push_config("features.remote_models=false");
    // Ambient plugins inject MCP servers and session-start hooks into PostHog
    // sessions; threads only get the MCP servers PostHog injects.
    push_config("features.plugins=false");
    // Model auth is injected via POSTHOG_GATEWAY_API_KEY, so codex's own
    // credential stores are unused: keep them on plain files, never a keychain.
    push_config("cli_auth_credentials_store=\"file\"");
    push_config("mcp_oauth_credentials_store=\"file\"");
    // No sandbox launcher on linux (a non-danger policy would panic); the
    // enclosing docker/Modal sandbox isolates instead.
    push_config("sandbox_mode=\"danger-full-access\"");

    if let Some(base_url) = options.api_base_url.as_deref().filter(|u| !u.is_empty()) {
        push_config("model_provider=\"posthog\"");
        push_config("model_providers.posthog.name=\"PostHog Gateway\"");
        push_config(&format!("model_providers.posthog.base_url=\"{base_url}\""));
        push_config("model_providers.posthog.wire_api=\"responses\"");
        push_config("model_providers.posthog.env_key=\"POSTHOG_GATEWAY_API_KEY\"");
    }

    args
}

pub struct SpawnedAppServer {
    pub child: Child,
    pub stdin: ChildStdin,
    pub stdout: ChildStdout,
}

pub fn spawn_app_server(options: &AppServerOptions) -> std::io::Result<SpawnedAppServer> {
    let args = build_app_server_args(options);
    let mut command = Command::new(&options.binary_path);
    command.args(&args);
    command.current_dir(&options.cwd);
    command.stdin(Stdio::piped());
    command.stdout(Stdio::piped());
    // codex diagnostics belong in /tmp/agent-server.log with ours.
    command.stderr(Stdio::inherit());
    command.kill_on_drop(true);

    if let Some(api_key) = options.api_key.as_deref().filter(|k| !k.is_empty()) {
        command.env("POSTHOG_GATEWAY_API_KEY", api_key);
    }
    if let Some(codex_home) = options.codex_home.as_deref().filter(|h| !h.is_empty()) {
        command.env("CODEX_HOME", codex_home);
    }

    let mut child = command.spawn()?;
    let stdin = child.stdin.take().expect("piped stdin");
    let stdout = child.stdout.take().expect("piped stdout");
    tracing::info!(
        binary = %options.binary_path.display(),
        pid = child.id(),
        "Spawned codex app-server"
    );
    Ok(SpawnedAppServer {
        child,
        stdin,
        stdout,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn args_pin_provider_and_credential_stores() {
        let options = AppServerOptions {
            binary_path: PathBuf::from("/usr/bin/codex"),
            cwd: "/tmp".into(),
            api_base_url: Some("https://gateway.example/v1".into()),
            api_key: Some("k".into()),
            codex_home: None,
        };
        let args = build_app_server_args(&options);
        assert_eq!(args[0], "app-server");
        let joined = args.join(" ");
        assert!(joined.contains("features.remote_models=false"));
        assert!(joined.contains("features.plugins=false"));
        assert!(joined.contains("cli_auth_credentials_store=\"file\""));
        assert!(joined.contains("sandbox_mode=\"danger-full-access\""));
        assert!(joined.contains("model_provider=\"posthog\""));
        assert!(joined.contains("model_providers.posthog.base_url=\"https://gateway.example/v1\""));
        assert!(joined.contains("model_providers.posthog.wire_api=\"responses\""));
        assert!(joined.contains("model_providers.posthog.env_key=\"POSTHOG_GATEWAY_API_KEY\""));
    }

    #[test]
    fn args_skip_provider_without_base_url() {
        let args = build_app_server_args(&AppServerOptions {
            binary_path: PathBuf::from("/usr/bin/codex"),
            cwd: "/tmp".into(),
            ..Default::default()
        });
        assert!(!args.join(" ").contains("model_provider"));
    }
}
